misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt

```diff
@@ -46,6 +53,188 @@ internal data class Variables(
+   * column to select rows from the table.
+   *
+   * If the value is func, the value used is the result of some function. To see which function,
+   * use SHOW WARNINGS following EXPLAIN to see the extended EXPLAIN output. The function might
+   * actually be an operator such as an arithmetic operator.
+   */
+  internal var ref: String? = null
+  /**
+   * The rows column indicates the number of rows MySQL believes it must examine to execute the
+   * query.
+   *
+   * For InnoDB tables, this number is an estimate, and may not always be exact.
+   */
+  internal var rows: BigInteger? = null
+  /**
+   * The filtered column indicates an estimated percentage of table rows that will be filtered by
+   * the table condition. That is, rows shows the estimated number of rows examined and
+   * rows × filtered / 100 shows the number of rows that will be joined with previous tables.
+   */
+  internal var filtered: Double? = null
+  /**
+   * This column contains additional information about how MySQL resolves the query. For
+   * descriptions of the different values, see
+   * https://dev.mysql.com/doc/refman/5.7/en/explain-output.html#explain-extra-information.
+   */
+  internal var Extra: String? = null
+
+  // MySQL chose a key, let's make sure it's in possible_keys. If not, the key isn't likely
+  // going to be helpful.
+  // Sometimes MySQL knows the answer without having to do any work.
+  fun isIndexed(): Boolean {
+    if (key != null && possible_keys != null) {
+      if (possible_keys!!.contains(key!!)) {
+        return true
+      }
+    }
+    return Extra != null && NO_INDEX_NEEDED_MESSAGES.contains(Extra!!)
+  }
+
+  /**
+   * If the query has possible keys, takes input joined from another subquery, or uses the
+   * PRIMARY key to fetch a limited number of rows, allow with a warning.
+   */
+  internal fun isProbablyOkay(query: String): Boolean {
+    if (possible_keys != null && !possible_keys!!.isEmpty()) {
+      return true
+    }
+
+    if (table != null) {
+      if (table!!.startsWith("<union")) {
+        // {code}<unionM,N>{code}: The row refers to the union of the rows with id values of M
+        // and N.
+        return true
+      }
+      if (table!!.startsWith("<derived")) {
+        // {code}<derivedN>{code}: The row refers to the derived table result for the row with an
+        // id value of N. A derived table may result, for example, from a subquery in the FROM
+        // clause.
+        return true
+      }
+      if (table!!.startsWith("<subquery")) {
+        // {code}<subqueryN>{code}: The row refers to the result of a materialized subquery for
+        // the row with an id value of N.
+        // See https://dev.mysql.com/doc/refman/5.7/en/subquery-materialization.html.
+        return true
+      }
+    }
+
+    return "PRIMARY" == key && query.contains(" limit ")
+  }
+
+  override fun toString(): String {
+    val builder = StringBuilder()
```

Comment:
You might like buildString.

Search the codebase for an example.

Reply:
Ohhhh coooool!